capture sources-state.json during checkout_layers

js731ca opened the pull request:
When using a layers-setup.conf.json that points only to branches or tags, e.g.:
```
    "bitbake-setup": {
<snip>
    "sources": {                                                                                                                            
        "bitbake": {                                                                                                                        
            "git-remote": {                                                                                                                 
                "branch": "2.8",                                                                                                            
                "describe": "",                                                                                                             
                "remotes": {                                                                                                                
                    "upstream": {                                                                                          
                        "uri": "https://github.com/openembedded/bitbake"                                                                
                    }                                                                                                                       
                },                                                                                                                          
                "rev": "2.8"                                                                                                                
            },                                                                                                                              
            "path": "bitbake"                                                                                                               
        }, 
<snip>
```
the actual revision cloned during a `bitbake-setup init` or `bitbake-setup update` is not recorded.

This changeset would extend the `checkout_layers` to return a sources datastructure, that is then put under version-control as `build/config/sources-state.json` - alongside the build/config/upstream-config.json 

The idea is that to reproduce any build, one could use this file, and feed it back into `bitbake-setup init --sources-overrides`

~A future(?) addition could be~ Adding also a new command that takes the upstream-config.json and the sources-state to write out a self-contained/complete layers-setup.conf.json. This file could then be archived in the registry.

~WIP because of some ToDos in the code that could use bitbake expert knowledge ;-)~

----                 
Example usecase/scenario:
```
# a CI is tasked to build a tip-of-master configuration once a day:
./bitbake/bin/bitbake-setup \
    --setting default top-dir-prefix /tmp/builds --setting default top-dir-name nightly \
    init --non-interactive acme_master.conf.json acme distro/acme machine/gizmo 

# then a build is run, its result is checked, tested, etc, etc

# once deemed OK, a snapshot is stored to later reproduce that exact build:
./bitbake/bin/bitbake-setup \
    --setting default top-dir-prefix /tmp/builds --setting default top-dir-name nightly \
    dump  \
    --build-dir /tmp/builds/nightly/acme_master_master-acme-distro_acme-machine_gizmo/ \
    --output registry/nightly/2025-11-01.conf.json
```

js731ca committed:
bitbake-setup: capture revisions while checking out layers

When initializing a build setup from a conf.json that only sets 'rev'
to a tag or branch, the actual revision would not be captured or
logged.

To capture the current layer state after an 'init' or 'update', the
checkout_layers function is extended to store the revision the
bb.fetch.Fetch pulled, and write that information into a
sources-state.json file. This file can then be fed back into
bitbake-setup init as: --sources-overrides

This new 'sources-state.json' is written during 'update_build' and
stored alongside the 'config-upstream.json' in the config dir. And put
with the later under version control by calling 'commit_config" after
'update_build'.

The use of 'deepcopy' is necessary to not modify the original input
data - which python passes around as reference.

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca committed:
bitbake-setup: rename 'write_config' into 'write_upstream_config'

Rename the function to reflect which (of the possibly many) files in
the build/config directory it writes to.

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca commented on bin/bitbake-setup:
```diff
@@ -98,12 +115,18 @@ def checkout_layers(layers, layerdir, d):
             do_fetch(fetcher, layerdir)
+            #TODO: there *has* to be a non-hacky way to get to the revision the fetcher just fetched?
+            k = list(fetcher.ud.keys())[0]
+            revision = fetcher.ud[k].revision
```
RFC: i haven't found a clean way yet... suggestions? :-)

kanavin replied:
Before you do any further work, try `bitbake-layers create-layers-setup` - achieves roughly same thing I think. Maybe that should be tweaked to better integrate with bitbake-setup.

js731ca replied:
thanks for the reminder/hint!
took another look: bitbake-layers and the setup-layers(.json) gets close, but:
- it only considers what is setup in the `bb-layers`, so `bitbake` itself is missing :-(
- it only covers layers of the current configuration: if a bitbake-setup has two configurations: gizmo: (meta-a, meta-b) and gadget: (meta-a, meta-c) - bitbake-layers can only capture one of them
- bitbake-layers only captures the 'sources' list, all other 'bitbake-setup' configurations are unknown to it

the current state already works - so you could give it a spin an see what files fall out :-)

kanavin replied:
Right, I just want to note that I'm on hoilday for the next week, and would rather not look at code until the 20th :) I wanted to get the settings UI rework in, as the matter was urgent but these things will have to wait a bit.

FWIW, I agree with the points, the use cases are perhaps different:

- bitbake-layers's use case is "I have an active yocto/bitbake build, can you create me a configuration that can be used to reproduce it". It needs to be extended to produce bitbake-setup configurations too, e.g. get the list of active fragments, figure out where bitbake is, ensure local.conf is empty, and then produce a fully usable json for bitbake-setup.

- bitbake-setup needs to capture the revisions of what it has cloned (or updated), at least to make a record of them under version control (some people previously asked me why do we need a version controlled history of configurations in the build dir, here's one reason why). That way a generally non-reproducible configuration can be used to produce reproducible ones with frozen layer revisions in specific invocations of bitbake-setup.

js731ca replied:
enjoy your vacation!
i'll go radio silent then - no new pushes or comments ;-)

kanavin commented:
I am back and bitbake-setup has meanwhile landed in master! So you can rebase/rework/etc the incomplete work as needed and I'll take a look.

js731ca commented:
8317a78 is just a rebase ontop the current bitbake:master

isn't bitbake-setup there to take over the setup-layers part?
setup-layers captures the layers (withouth BB) and creates a bootstrap script to (try to) reproduce the build setup - which is IMO now better covered by bitbake-setup :-)

back to the RFC: do you know a non-hacky way to get the git-hash from the bitbake.fetcher after it cloned a repository?

kanavin commented:
> isn't bitbake-setup there to take over the setup-layers part? setup-layers captures the layers (withouth BB) and creates a bootstrap script to (try to) reproduce the build setup - which is IMO now better covered by bitbake-setup :-)

setup-layers is a kind of precursor to bitbake-setup. We didn't immediately have a clear idea of how this should work, and experimented with different approaches. setup-layers also has the benefit of being a standalone python script that doesn't require cloning bitbake and satisfying its host requirements.
 
> back to the RFC: do you know a non-hacky way to get the git-hash from the bitbake.fetcher after it cloned a repository?

I couldn't find it in the fetcher either. You need to run 'git rev-parse HEAD' after each unpack operation.

js731ca committed:
bitbake-setup: checkout_layers: construct 'src_uri' separatly

Construct the 'src_uri' separately, and then pass either variant into
one call that creates the Fetch. Making use of format-strings to
shorten/simplify the code.

Also: using 'proto' instead of 'type' for a variable name, to avoid
the protected keyword.

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca commented:
> > isn't bitbake-setup there to take over the setup-layers part? setup-layers captures the layers (withouth BB) and creates a bootstrap script to (try to) reproduce the build setup - which is IMO now better covered by bitbake-setup :-)
> 
> setup-layers is a kind of precursor to bitbake-setup. We didn't immediately have a clear idea of how this should work, and experimented with different approaches. setup-layers also has the benefit of being a standalone python script that doesn't require cloning bitbake and satisfying its host requirements.
> 

that benefit can also backfire: our project generated once a setup-layers script, and then started to extend/maintain it downstream to also: clone/setup layers/bitbake, add an option to 'skip-revs' = checkout the tip of each branch instead of the pinned revision, ...

> > back to the RFC: do you know a non-hacky way to get the git-hash from the bitbake.fetcher after it cloned a repository?
> 
> I couldn't find it in the fetcher either. You need to run 'git rev-parse HEAD' after each unpack operation.

bitbake/lib/bb/fetch2/git.py already does this - I've pondered refactoring the fetch2/git.py to have an accessor-function - but the whole python module is geared a bit too much to bitbake recipe parsing :-|
But after restructuring the bitbake-setup code a bit, the "hack" in c9536ed now seems less... "bad".
since we setup a Fetch with a known git src_uri, we can assume a thing or two, and then get the revision directly from the `urldata` inside the Fetch.

js731ca commented:
v3: open for review, since it now works as intended :-)
try this:
- bitbake-setup init, with a master.conf.json (a configuration that points to rev=HEAD/master/main)
- bitbake-setup init --source-overrides build-dir/conf/sources-states.json master.conf.json = reuse the previously recorded revisions
- bitbake-setup dump --build-dir build-dir/acme-gizmo --output gizmo_2025-11-01.conf.json = get a complete bitbake.conf.json back out
- bitbake-setup init gizmo_2025-11-01.conf.json = reproduce the previous build environment, down to the revisions

kanavin commented:
Thanks for the updates! I'm currently swamped with getting bitbake-setup docs done (at least the most critical parts), but will certainly get to this. It's not forgotten.

js731ca commented:
> ... getting bitbake-setup docs done

want/need a proof reader who already played with the code? :-)

kanavin commented:
> > ... getting bitbake-setup docs done
> 
> want/need a proof reader who already played with the code? :-)

You can find the patches and reviews on bitbake-devel and yocto-docs :) Antonin is the principal writer, I'm reviewing and adding details.

kanavin commented on bin/bitbake-setup:
```diff
@@ -471,8 +487,8 @@ def init_config(top_dir, settings, args, d):
     write_upstream_config(confdir, upstream_config)
-    commit_config(confdir)
     update_build(upstream_config, confdir, builddir, layerdir, d)
+    commit_config(confdir)
```
This change needs a mention in the commit message.

js731ca replied:
ACK
> This new 'sources-state.json' is written during 'update_build' and stored alongside the 'config-upstream.json' in the config dir. And put with the later under version control by calling 'commit_config" after 'update_build'.

kanavin commented on bin/bitbake-setup:
```diff
@@ -269,15 +283,17 @@ def get_registry_config(registry_path, id):
 
 def update_build(config, confdir, builddir, layerdir, d):
-    layer_config = config["data"]["sources"]
+    layer_config = copy.deepcopy(config["data"]["sources"])
```
This is probably unneeded? If it is needed or beneficial, a comment in the code would be helpful.

js731ca replied:
it is - python passes those data structures around as reference; and without the deepcopy we end up modifying the original -> not what we want here

kanavin replied:
But where is it modified? I am not seeing that?

js731ca replied:
a few lines further down, when the layer_overrides are applied

kanavin replied:
Ah. But in that case, we need to capture the *original* overrides as well as the final fixed revisions set!

kanavin replied:
Nevermind, I'm confused :)

kanavin commented on bin/bitbake-setup:
```diff
@@ -58,6 +58,15 @@ def get_config_name(config):
+    filename = os.path.join(config_dir, config_name)
+    if not os.path.exists(filename):
+        print(f"Warning: '{filename}' does not exist, returning empty set")
+        return {}
```
I don't understand, should never happen to begin with, and if it does happen, then this should be a hard error? What prompted you to make this commit?

js731ca replied:
the `read_config` function is also used to read the new sources-state.json, which might not yet exist if a user comes from an older version of bitbake-setup -- that's where the defensiveness came from

kanavin replied:
Since we're dropping the 'dump' function for now, I think this can be removed as well.

js731ca replied:
ACK

kanavin commented on bin/bitbake-setup:
```diff
@@ -99,6 +107,10 @@ def checkout_layers(layers, layerdir, d):
             do_fetch(fetcher, layerdir)
+            urldata = fetcher.ud[src_uri]
+            revision = urldata.revision
+            print(f"      cloned: {fetchuri}\n      at revision: {revision}")
```
Can you show the sample output before and after this change?

js731ca replied:
Before:
```
Initializing a setup directory in
    /tmp/bitbake-setup/gs/geosurv_master-geosurv-distro_geosurv-machine_beluga
Fetching layer/tool repositories into /tmp/bitbake-setup/gs/geosurv_master-geosurv-distro_geosurv-machine_beluga/layers
    bitbake
    meta-arm      
<snip>
```
and After:
```
Initializing a setup directory in
    /tmp/bitbake-setup/gs/geosurv_master-geosurv-distro_geosurv-machine_beluga
Fetching layer/tool repositories into /tmp/bitbake-setup/gs/geosurv_master-geosurv-distro_geosurv-machine_beluga/layers
    bitbake
      cloned: git://github.com/hexagon-geo-surv/bitbake
      at revision: 35a421cdcafb3595b9de5489ffdc567825400d26
    meta-arm
      cloned: git://github.com/hexagon-geo-surv/meta-arm
      at revision: b50b4f31e668754ae8b887e711126f5b92363161
<snip>
```

kanavin replied:
I'm not sure this is necessarily an improvement. Can you split this into a separate pull request?

js731ca replied:
i could/can, one attempt to sell the feature, and keep it here:
IMO its beneficial for cases where one only has the (CI) logs available to figure out why a build failed; considering a transient build setup (= the files are already gone)

kanavin replied:
I'd claim that a CI pipeline that preserves only the console output, and not any files from the build tree is broken in itself. Otherwise we're going down a rabbit hole of adding prints for this and that and this other thing, just in case any of it may help debugging a build fail.

In any case, this is an unrelated change, and shouldn't be bundled with the main feature.

kanavin replied:
That said, it can perhaps still be done, just not by default and subject to some kind of --verbose option.

kanavin commented on bin/bitbake-setup:
```diff
@@ -803,6 +819,12 @@ def main():
+    parser_dump.add_argument('--output', action='store', default='build-dump.conf.json', metavar='FILENAME',
+                             help="A filename to dump the bitbaek-setup conf.json data into. Defaults to 'build-dump.conf.json'")
+    parser_dump.set_defaults(func=build_dump)
+
```
This is where the patchset enters the 'tricky' territory. The problem with dumping the original registry config with particular revisions like that is that it does not capture the selections the user made, and so cannot be truly used to reproduce a particular build, as the users will have to answer chose-one-of questions without knowing the original answers. Some users may also not want an 'integrated' config like that and would prefer to have config+overrides.

I'd say rather the original config should be captured in the config/ repo as a standalone file, and then this new command isn't needed, as you already have the usable original config and overrides available by default, and you can just copy them out from the config/ repo. Then what kind of tooling should we have to synthesize further variations from everything that was captured in config/ is a separate question, that can be addressed in separate pull requests.

js731ca replied:
good point - i'm still thinking too much along the lines of setup-layers, and am still working with a fragment-less bitbake-setup config = haven't considered the fragments and user choices.

the usage scenario i had in mind for this is:
1. a ci-setup always would call a one-liner with a known selection: bitbake-setup init --non-interactive \<snip> acme-master.conf.json \<snip>
2. build, test, ...
3. bitbake-setup dump -> commit to registry
4. a user would (have to) take the same "one-liner with a known selection" from 1., but now could use the dump-ed conf.json instead of acme-master.conf.json

so capturing the 'sources-state.json' would be enough - as it can be directly plugged back in as sources-overrides; and i should just drop the commits for the 'dump' command?

kanavin replied:
Basically, yes, you should drop the last two 'dump' commits, and then I'm more or less okay with the set. This can be followed later by further tweaks or refinements.

js731ca commented:
v4: rebase onto bitbake:master, dropping
- defensive "read_config" (  9986ca7425c66c9ced88bfaeac8c6cace71bda2e ) and
- the "dump" command ( f7b53dbd1bb2a87b9eadb4fcfe061f973adfde99 + f724371d3fca7fa28e1048226273c612c826020b )

kanavin commented:
> v4: rebase onto bitbake:master, dropping
> 
>     * defensive "read_config" (  [9986ca7](https://github.com/kanavin/bitbake/commit/9986ca7425c66c9ced88bfaeac8c6cace71bda2e) ) and
> 
>     * the "dump" command ( [f7b53db](https://github.com/kanavin/bitbake/commit/f7b53dbd1bb2a87b9eadb4fcfe061f973adfde99) + [f724371](https://github.com/kanavin/bitbake/commit/f724371d3fca7fa28e1048226273c612c826020b) )

You need to rebase again. I only pushed newly added commits from upstream master (no force pushing), and this somehow *still* confused github.

kanavin commented on bin/bitbake-setup:
```diff
@@ -61,6 +62,12 @@ def write_upstream_config(config_dir, config_data):
+def write_sources_state(config_dir, config_data):
+    sources = {}
+    sources['sources'] = config_data
+    with open(os.path.join(config_dir, "sources-state.json"),'w') as s:
```
I think this (and the function name) should be named 'sources-fixed-revisions.json', which is more clear.

kanavin commented on bin/bitbake-setup:
```diff
@@ -76,6 +83,7 @@ def _write_layer_list(dest, repodirs):
         json.dump({"version":"1.0","layers":layers}, f, sort_keys=True, indent=4)
 
 def checkout_layers(layers, layerdir, d):
+    layer_state = copy.deepcopy(layers)
```
Same here, layers_fixed_revisions is better.

kanavin commented on bin/bitbake-setup:
```diff
@@ -61,6 +62,12 @@ def write_upstream_config(config_dir, config_data):
+    sources = {}
+    sources['sources'] = config_data
+    with open(os.path.join(config_dir, "sources-state.json"),'w') as s:
+        json.dump(sources, s, sort_keys=True, indent=4)
```
This needs an update to bitbake-setlftest as well: that the file exists, and contains revisions that match the current layer checkout.

js731ca committed:
bitbake-setup: tests: move duplicate code into 'check_setupdir_files'

All places that called into 'check_setupdir_files' did the same
preparation step to load the upstream-config.json and then pass it
into the function.

Since the 'setuppath' is already passed into the function, and the
name and relative location of the upstream-config.json is fixed,
constructing the file path and loading the json could be done in the
function.

De-duplicate code by loading the json inside the function instead.

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca commented:
v5:
- renames functions and variables to the recommended 'sources-fixes-revisions'
- drops the added prints
- adds a test (and refactors the tests a little bit)
